src/backend/iohidmanager/device.rs

```diff
@@ -117,38 +117,50 @@ impl IOHIDDevice {
     }
 
     pub fn register_input_report_callback<F>(&self, callback: F) -> HidResult<CallbackGuard>
-    where
-        F: FnMut(&[u8]) + Send + 'static
+        where
+            F: FnMut(&[u8]) + Send + Sync + 'static
     {
         let max_input_report_len = self.get_i32_property(kIOHIDMaxInputReportSizeKey)? as usize;
 
-        let mut report_buffer = vec![0u8; max_input_report_len].into_boxed_slice();
+        let mut report_buffer = ManuallyDrop::new(vec![0u8; max_input_report_len]);
+
+        let (report_buffer_ptr, report_buffer_len, report_buffer_capacity) = (report_buffer.as_mut_ptr() as usize, report_buffer.len(), report_buffer.capacity());
+
         let callback: InputReportCallback = Box::new(callback);
-        let callback = Box::new(UnsafeCell::new(callback));
+        let callback: InputReportCallbackContainer = Box::new(UnsafeCell::new(callback));
+
         unsafe {
             IOHIDDeviceRegisterInputReportCallback(
                 self.as_concrete_TypeRef(),
-                report_buffer.as_mut_ptr(),
-                report_buffer.len() as _,
+                report_buffer_ptr as _,
+                report_buffer_len as _,
                 hid_report_callback,
                 callback.get() as _
             );
         }
+
+        let callback_ptr = Box::into_raw(callback) as usize;
+
         Ok(CallbackGuard {
             device: self.clone(),
-            _report_buffer: report_buffer,
-            _callback: callback
+            report_buffer_ptr,
+            report_buffer_len,
+            report_buffer_capacity,
+            callback_ptr,
         })
     }
 }
 
-type InputReportCallback = Box<dyn FnMut(&[u8]) + Send>;
+type InputReportCallback = Box<dyn FnMut(&[u8]) + Send + Sync>;
+type InputReportCallbackContainer = Box<UnsafeCell<InputReportCallback>>;
```

Comment:
I think I overthought things when I wrote this and we don't need the `UnsafeCell`, especially not the way I used it.